tfc-dump.pl

```diff
@@ -30,140 +58,193 @@
-	$curl_query = "\"https://app.terraform.io/api/v2/workspaces/$workspace_list{$_}\"";
-	$curl_cmd   = "curl $curl_headers --output $_-attributes.json $curl_query";
-	system($curl_cmd);
-
-	# Dump the variables info
-	$curl_query = "\"https://app.terraform.io/api/v2/workspaces/$workspace_list{$_}/vars\"";
-	$curl_cmd   = "curl $curl_headers --output $_-variables.json $curl_query";
-	system($curl_cmd);
+    # Dump the workspace info
+    $curl_query = "\"https://app.terraform.io/api/v2/workspaces/$workspace_list{$_}\"";
+    $curl_cmd   = "curl $curl_headers --output $_-attributes.json $curl_query";
+    if (system($curl_cmd) != 0) {
+        my $error = "Failed to dump workspace attributes for $_";
+        error_to_sentry($error);
+        print STDERR "$error\n";
+    }
+
+    # Dump the variables info
+    $curl_query = "\"https://app.terraform.io/api/v2/workspaces/$workspace_list{$_}/vars\"";
+    $curl_cmd   = "curl $curl_headers --output $_-variables.json $curl_query";
+    if (system($curl_cmd) != 0) {
+        my $error = "Failed to dump workspace variables for $_";
+        error_to_sentry($error);
+        print STDERR "$error\n";
+    }
 }
 
 # Dump the variable sets data to files.
-
-### WARNING ###
-#
-# This code assumes that all of the TFC Variable Sets are contained within
-# the first result page of 100 entries.  This was true for SIL in August 2024.
-#
-####
-
 my @vs_names;
 my @vs_ids;
-my $pg_size = 100;
+my $pg_size = 20;  # Default page size per Hashicorp documentation
+my $pg_num = 1;    # Start with page 1
 my $total_count;
-my $tmpfile = `mktemp`;
+my $total_processed = 0;
+my $tmpfile;
+system("mktemp > /tmp/tmpfile.txt");
+$tmpfile = do {
+    local(@ARGV, $/) = '/tmp/tmpfile.txt';
+    <>;
+};
+system("rm -f /tmp/tmpfile.txt");
 chomp($tmpfile);
 
-$curl_query = "\"https://app.terraform.io/api/v2/organizations/${tfc_org_name}/varsets?page%5Bsize%5D=${pg_size}\"";
-$curl_cmd   = "curl $curl_headers --output $tmpfile $curl_query";
-system($curl_cmd);
-
-# Get the Variable Set names
-
-$jq_cmd   = "cat $tmpfile | jq '.data[].attributes.name'";
-@vs_names = `$jq_cmd`;
-# Remove the double quotes in each element of the array.
-grep($_ =~ s/"//g && 0, @vs_names);	# Programming Perl, p. 221, 1990
-chomp(@vs_names);
-
-# Get the Variable Set IDs
+do {
+    # Get the current page of variable sets
+    $curl_query = "\"https://app.terraform.io/api/v2/organizations/${tfc_org_name}/varsets?page%5Bsize%5D=${pg_size}&page%5Bnumber%5D=${pg_num}\"";
+    $curl_cmd = "curl $curl_headers --output $tmpfile $curl_query";
+    
+    if (system($curl_cmd) != 0) {
+        my $error = "Failed to fetch variable sets page $pg_num";
+        error_to_sentry($error);
+        print STDERR "$error\n";
+        unlink($tmpfile);
+        exit(1);
+    }
+
+    # Get the Variable Set names for this page
+    $jq_cmd = "cat $tmpfile | jq '.data[].attributes.name'";
+    system("$jq_cmd > /tmp/page_names.txt");
+    my @page_names = do {
```

Comment:
... then you could slurp_output an array for here, and string elsewhere.

Reply:
Or rename slurp_output to slurp_string_output and slurp_array_output, and extract the first overlapping part to slurp_to_file. That way each function only returns one type (which is nicer coding).